File: src/main/java/com/derongan/minecraft/guiy/gui/ClickableElement.java
package com.derongan.minecraft.guiy.gui;

import java.util.function.Consumer;

/**
 * Wrapper {@link Element} with a click callback.
 */
public class ClickableElement implements Element {
    private Element wrapped;
    private Consumer<ClickEvent> pickupAction;

    /**
     * @param wrapped Element to delegate all clicks to.
     */
    public ClickableElement(Element wrapped) {
        this.wrapped = wrapped;
        this.pickupAction = a -> a.setCancelled(true);
    }

    @Override
    public Size getSize() {
        return wrapped.getSize();
    }

    @Override
    public void draw(GuiRenderer guiRenderer) {
        wrapped.draw(guiRenderer);
    }

    @Override
    public void onClick(ClickEvent clickEvent) {
        pickupAction.accept(clickEvent);
    }

    /**
     * Sets the onClick action.
     *
     * @param clickAction the action to take when clicked.
     */
    public void setClickAction(Consumer<ClickEvent> clickAction) {
        this.pickupAction = clickAction;
    }
}

File: src/main/java/com/derongan/minecraft/guiy/gui/ClickableElement.kt
package com.derongan.minecraft.guiy.gui

import java.util.function.Consumer


/**
 * Wrapper [Element] with a click callback.
 *
 * @param wrapped Element to delegate all clicks to.
 * @param clickAction the action to take when clicked.
 */
open class ClickableElement @JvmOverloads constructor(
        private val wrapped: Element,
        var clickAction: (ClickEvent) -> (Unit) = { it.isCancelled = true }
) : Element {
    override val dims: Size
        get() = wrapped.dims

    override fun draw(guiRenderer: GuiRenderer) {
        wrapped.draw(guiRenderer)
    }

    override fun onClick(clickEvent: ClickEvent) {
        clickAction(clickEvent)
    }

    /**
     * Sets the onClick action.
     * Only visible to java.
     *
     * @param clickAction the action to take when clicked.
     */
    @Suppress("NEWER_VERSION_IN_SINCE_KOTLIN")
    @kotlin.SinceKotlin(version = "99999.0") //hack to prevent viewing in Kotlin
    fun setClickAction(clickAction: Consumer<ClickEvent>) {
        this.clickAction = clickAction::accept
    }
}